checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -363,57 +428,132 @@ private static int getGroupCount(Pattern p) {
     }
 
     /**
-     * Return the non-null groups in the argument.
+     * Returns a list of groups other than 0, that are guaranteed to be non-null given that the
+     * regular expression matches the String. The String argument passed has to be a valid regular
+     * expression.
      *
-     * @param p pattern to be analysed
-     * @return the groups that are non-null in the argument
+     * @param regexp pattern to be analysed
+     * @param n number of capturing groups in the pattern
+     * @return a {@code List} of groups that are guaranteed to match some part of a string that
+     *     matches {@code regexp}
+     * @throws Error if the argument is not a regex
      */
-    private static List<Integer> getNonNullGroups(Pattern p) {
-        String regexp = p.pattern();
-        int n = getGroupCount(p);
+    public static List<Integer> getNonNullGroups(String regexp, int n) {
+        try {
+            Pattern.compile(regexp);
+        } catch (PatternSyntaxException e) {
+            throw new Error(e);
+        }
         List<Integer> nonNullGroups = new ArrayList<>();
-        for (int i = 0; i <= n; i++) {
+        for (int i = 1; i <= n; i++) {
             nonNullGroups.add(i);
         }
+
+        // Holds all indices of opening parentheses that were openings of a capturing group.
         ArrayDeque<Integer> openingIndices = new ArrayDeque<>();
-        int group = 0;
-        boolean squareBracketOpen = false;
+        // Holds whether the last occurrence of a non-literal '(' was a capturing group (true)
+        // or was some other special construct (false). This helps in determining whether a
+        // non-literal
+        // ')' closes a capturing group or some other special construct. If it closes a capturing
+        // group, the top element from openingIndices has to be removed.
+        ArrayDeque<Boolean> openingWasGroup = new ArrayDeque<>();
+        // If true, the character just before the current one was '\', i.e. the current character
+        // has to be considered either in a literal sense or as some special character or flag.
         boolean escaped = false;
-        boolean notCapturing = false;
-        for (int i = 0; i < regexp.length(); i++) {
-            if (!escaped && !squareBracketOpen && regexp.charAt(i) == '(') {
-                if (i != regexp.length() - 1 && regexp.charAt(i + 1) == '?') {
-                    notCapturing = true;
-                    continue;
+        int group = 0;
+
+        /**
+         * If you encounter '(' and it is not preceded by a '\', then it is a special group. If it
```

Comment:
By "special group" I think you mean the same concept as the Pattern Javadoc calls "pure, non-capturing groups that do not capture text".  It is clearer to use the same terminology that the JDK uses.